libraries/commerce/favorites/actions/toggleFavorites.js

```diff
@@ -5,90 +5,117 @@ import * as pipelines from '../constants/Pipelines';
   const count = getFavoritesCount(getState());
   if (count > limit) {
     const error = new Error('Limit exceeded');
     error.code = ELIMIT;
     dispatch(errorFavorites(productId, error));
+  }
+};
+/**
+ * Calls Add Favorites pipeline
+ * @param {string} productId Product identifier
+ * @returns {Promise} PipelineRequest dispatch.
+ */
+const syncAddSingleFavorite = productId => (
+  new PipelineRequest(pipelines.SHOPGATE_USER_ADD_FAVORITES)
+    .setInput({ productId })
+    .setRetries(0)
+    .dispatch()
+);
+
+/**
+ * Calls Add Favorites pipeline
+ * @param {string} productId Product identifier
+ * @returns {Promise} PipelineRequest dispatch.
+ */
+const syncRemoveSingleFavorite = productId => (
+  new PipelineRequest(pipelines.SHOPGATE_USER_DELETE_FAVORITES)
+    .setInput({ productId })
+    .setRetries(0)
+    .dispatch()
+);
+/**
+ * Sync Add Favorites to backend
+ * @param {[Object]} bufferedActions Actions that have been buffered
+ * @return {Function}
+ */
+const syncAddRemoveFavorites = bufferedActions => (dispatch) => {
+  const uniqueProductIds = [...new Set(bufferedActions
+    .filter(action => !!action.productId)
+    .map(action => action.productId))];
+  const productIdsToAdd = [];
+  const productIdsToRemove = [];
+  uniqueProductIds.forEach((productId) => {
+    const addActions = bufferedActions.filter(action => (
+      productId === action.productId && REQUEST_ADD_FAVORITES === action.type
+    ));
+    const removeActions = bufferedActions.filter(action => (
+      productId === action.productId && REQUEST_REMOVE_FAVORITES === action.type
+    ));
+    // push product id to productIdsToAdd array if there are more
+    // add actions than remove actions for this product
+    if (addActions.length > removeActions.length) {
+      productIdsToAdd.push(productId);
+    }
+    // push product id to productIdsToRemove array if there are more
+    // remove actions than add actions for this product
+    if (removeActions.length > addActions.length) {
+      productIdsToRemove.push(productId);
+    }
+    // product id is not added to either productIdsToAdd or productIdsToRemove
+    // when there are as many add actions as remove actions
+  });
+  dispatch(requestAddRemoveFavoritesSync(productIdsToAdd, productIdsToRemove));
+  if (productIdsToAdd.length < 1 && productIdsToRemove.length < 1) {
+    // since there are no product to add or remove bale out
+    dispatch(idleSyncFavorites());
     return;
   }
 
-  syncPendingCount += 1;
-  dispatch(requestSync());
+  const addPromises = productIdsToAdd.map(productId => syncAddSingleFavorite(productId));
+  const removePromises = productIdsToRemove
+    .map(productId => syncRemoveSingleFavorite(productId));
+  Promise.all([...addPromises, ...removePromises])
+    .then(() => {
+      dispatch(receiveAddRemoveFavoritesSync(productIdsToAdd, productIdsToRemove));
+    }).catch((error) => {
+      dispatch(errorAddRemoveFavoritesSync(productIdsToAdd, productIdsToRemove, error));
```

Comment:
This will stop waiting for promises on the first error. That means it will report that error and not dispatch any "receive"  actions for successful requests that come in after the first error.
This could occur when a bunch of actions are buffered, and one of the early ones fails. In this case the frontend's state might become inconsistent (not displaying favorites, when they are added successfully, or still showing, when successfully removed).
Is this case already handled somewhere (requesting favorites a bit later or so)?

Reply:
It is handled like it was handled when the put pipeline was being used. If there is an error the lastChanged value of the favorites redux state object is set to 0 to `Reset both to fetch ASAP`. I took this comment to mean that there would be another get favorites call done which would ensure the favorites state is correctly synced to the backend.